airflow/models/dag.py

```diff
@@ -2991,14 +2991,38 @@ def deactivate_deleted_dags(cls, alive_dag_filelocs: List[str], session=NEW_SESS
                 continue
 
     @classmethod
-    def dags_needing_dagruns(cls, session: Session):
+    def dags_needing_dagruns(cls, session: Session) -> Tuple[Query, Dict]:
         """
         Return (and lock) a list of Dag objects that are due to create a new DagRun.
 
         This will return a resultset of rows  that is row-level-locked with a "SELECT ... FOR UPDATE" query,
         you should ensure that any scheduling decisions are made in a single transaction -- as soon as the
         transaction is committed it will be unlocked.
         """
+        from airflow.models.dataset import DatasetDagRef, DatasetDagRunQueue as DDRQ
+
+        # these dag ids are triggered by datasets, and they are ready to go.
+        dataset_triggered_dag_info_list = {
+            x.dag_id: (x.first_event_time, x.last_event_time)
+            for x in session.query(
+                DatasetDagRef.dag_id,
+                func.max(DDRQ.created_at).label('last_event_time'),
+                func.max(DDRQ.created_at).label('first_event_time'),
+            )
+            .join(
+                DDRQ,
+                and_(
+                    DDRQ.dataset_id == DatasetDagRef.dataset_id,
+                    DDRQ.target_dag_id == DatasetDagRef.dag_id,
+                ),
+                isouter=True,
+            )
```

Comment:
We should define the relationships on the DDRQ table (that way we only have to have it once, not in every place we might want to use this rel) then this can become something like


```suggestion
            .join(DatassetDagRef._something_)
```